File: frontend/src/types/timeline.ts
/**
 * Match-V5 timeline, re-keyed to PUUID by the backend so it joins straight onto the
 * scoreboard in `types/match.ts`.
 * @see GET /api/v1/matches/{matchId}/timeline
 */

/** A point in Riot's map coordinate space, not pixels. See `lib/timeline.ts`. */
export interface TimelinePosition {
  readonly x: number;
  readonly y: number;
}

/** Extent of the playable map, used to project a position onto the minimap image. */
export interface MapBounds {
  readonly min_x: number;
  readonly min_y: number;
  readonly max_x: number;
  readonly max_y: number;
}

/** One player's state at one frame. Riot samples these once a minute. */
export interface TimelineParticipantFrame {
  readonly puuid: string;
  readonly position: TimelinePosition;
  readonly level: number;
  readonly damage_to_champions: number;
  readonly xp: number;
  readonly cs: number;
  readonly current_gold: number;
  readonly total_gold: number;
  readonly health: number;
  readonly health_max: number;
  readonly armor: number;
  readonly magic_resist: number;
  readonly attack_damage: number;
  readonly ability_power: number;
  readonly movement_speed: number;
}

export interface TimelineFrame {
  readonly timestamp_ms: number;
  readonly participants: readonly TimelineParticipantFrame[];
}

export type TimelineEventType =
  | "CHAMPION_KILL"
  | "BUILDING_KILL"
  | "ELITE_MONSTER_KILL"
  | "WARD_PLACED"
  | "SKILL_LEVEL_UP"
  | "LEVEL_UP"
  | "ITEM_PURCHASED"
  | "ITEM_SOLD"
  | "ITEM_DESTROYED"
  | "ITEM_UNDO";

/** Fields not carried by a given event type are null. */
export interface TimelineEvent {
  readonly timestamp_ms: number;
  readonly type: TimelineEventType | string;
  readonly position: TimelinePosition | null;
  /** Killer, ward placer or item buyer. */
  readonly actor_puuid: string | null;
  readonly victim_puuid: string | null;
  readonly assist_puuids: readonly string[];
  readonly team_id: number | null;
  readonly item_id: number | null;
  /** 1-4 for Q, W, E, R. */
  readonly skill_slot: number | null;
  readonly level: number | null;
  readonly monster_type: string | null;
  readonly building_type: string | null;
  readonly lane_type: string | null;
  readonly ward_type: string | null;
}

export interface TimelineParticipantSummary {
  readonly puuid: string;
  /** Ground covered over the whole game, in Riot map units. */
  readonly distance_travelled: number;
}

export interface MatchTimeline {
  readonly match_id: string;
  readonly frame_interval_ms: number;
  readonly game_duration_ms: number;
  readonly map_id: number;
  readonly map_bounds: MapBounds;
  readonly participants: readonly TimelineParticipantSummary[];
  readonly frames: readonly TimelineFrame[];
  readonly events: readonly TimelineEvent[];
}

File: backend/app/schemas/timeline.py
"""Response shapes for the Match-V5 timeline.

Riot keys the timeline by `participantId` (1-10); everything here is re-keyed to PUUID
so the client can join it against the scoreboard it already has without carrying a
second identifier around.
"""

from pydantic import BaseModel, Field


class TimelinePosition(BaseModel):
    """A point in Riot's map coordinate space, not in pixels."""

    x: int
    y: int


class MapBounds(BaseModel):
    """Extent of the playable map, so the client can project positions onto its image.

    Riot's Y axis points up and a screen's points down, which is the client's job to
    flip. These are the raw game bounds.
    """

    min_x: int
    min_y: int
    max_x: int
    max_y: int


class TimelineParticipantFrame(BaseModel):
    """One player's state at one frame. Riot samples these once a minute."""

    puuid: str
    position: TimelinePosition
    level: int
    damage_to_champions: int = 0
    xp: int
    cs: int
    current_gold: int
    total_gold: int
    health: int
    health_max: int
    armor: int
    magic_resist: int
    attack_damage: int
    ability_power: int
    movement_speed: int


class TimelineFrame(BaseModel):
    timestamp_ms: int
    participants: list[TimelineParticipantFrame]


class TimelineEvent(BaseModel):
    """A notable moment. Fields not carried by a given event type stay None."""

    timestamp_ms: int
    type: str
    position: TimelinePosition | None = None
    actor_puuid: str | None = Field(
        default=None, description="Killer, ward placer or item buyer"
    )
    victim_puuid: str | None = None
    assist_puuids: list[str] = Field(default_factory=list)
    team_id: int | None = None
    item_id: int | None = None
    skill_slot: int | None = Field(default=None, description="1-4 for Q, W, E, R")
    level: int | None = None
    monster_type: str | None = None
    building_type: str | None = None
    lane_type: str | None = None
    ward_type: str | None = None


class TimelineParticipant(BaseModel):
    """Distance covered over the whole game, from the per-frame positions."""

    puuid: str
    distance_travelled: float


class MatchTimelineResponse(BaseModel):
    match_id: str
    frame_interval_ms: int
    game_duration_ms: int
    map_id: int
    map_bounds: MapBounds
    participants: list[TimelineParticipant]
    frames: list[TimelineFrame]
    events: list[TimelineEvent]
